packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -0,0 +1,384 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	type FieldKey,
+	type FieldKindIdentifier,
+	LeafNodeStoredSchema,
+	MapNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	type TreeFieldStoredSchema,
+	type TreeNodeSchemaIdentifier,
+	type TreeStoredSchema,
+	type TreeTypeSet,
+	type ValueSchema,
+} from "../../core/index.js";
+
+/**
+ * @remarks
+ *
+ * 1. FieldIncompatibility
+ *
+ * `FieldIncompatibility` represents the differences between two `TreeFieldStoredSchema` objects. It consists of
+ * three types of incompatibilities:
+ *
+ * - FieldKindIncompatibility: Indicates the differences in `FieldKindIdentifier` between two `TreeFieldStoredSchema`
+ * objects (e.g., optional, required, sequence, etc.).
+ * - AllowedTypesIncompatibility: Indicates the differences in the allowed child types between the two schemas.
+ * - ValueSchemaIncompatibility: Specifically indicates the differences in the `ValueSchema` of two
+ * `LeafNodeStoredSchema` objects.
+ *
+ * When comparing two `TreeFieldStoredSchema` objects, the `trackFieldDiscrepancies` function first identifies
+ * `AllowedTypesIncompatibility`, followed by `FieldKindIncompatibility`.
+ *
+ * 2. NodeIncompatibility
+ *
+ * `NodeIncompatibility` represents the differences between two `TreeNodeStoredSchema` objects and includes:
+ *
+ * - NodeKindIncompatibility: Indicates the differences in the types of `TreeNodeStoredSchema` (currently supports
+ * `ObjectNodeStoredSchema`, `MapNodeStoredSchema`, and `LeafNodeStoredSchema`).
+ * - NodeFieldsIncompatibility: Indicates the `FieldIncompatibility` of `TreeFieldStoredSchema` within two
+ * `TreeNodeStoredSchema`. It includes an array of `FieldIncompatibility` instances in the `differences` field.
+ *
+ * In the `getAllowedContentIncompatibilities` function, `NodeKindIncompatibility` is identified first. If the
+ * node kinds are consistent, the function then proceeds to find `NodeFieldsIncompatibility`.
```

Comment:
I'd phrase this as

```
// When comparing two nodes for compatibility, it only makes sense to compare their fields if the nodes are of the same kind (map, object, leaf).
```

rather than reference the specific (non-exported) function.